fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/FDBJoinIndexTest.java

```diff
@@ -0,0 +1,286 @@
+                        .setRecNo(parentDoc)
+                        .setIntValue(parentDoc * 100 + 3)
+                        .setTextValue("Here is my text #" + group + "/" + parentDoc)
+                        .setChildRecNo(-1 * parentDoc)
+                        .build();
+                if (!data.missingParent(parentRecord)) {
+                    data.add(parentRecord, recordStore.saveRecord(parentRecord));
+                }
+                if (!data.missingChild(parentRecord)) {
+                    var childRecord = TestRecordsGroupedParentChildProto.MyChildRecord.newBuilder()
+                            .setGroup(group)
+                            .setRecNo(-1 * parentDoc)
+                            .setOtherValue(parentDoc * 1000 + 52)
+                            .setStrValue("Children " + group + " grouping " + parentDoc)
+                            .build();
+                    data.add(childRecord, recordStore.saveRecord(childRecord));
+                }
+            }
+        }
+        return data;
+    }
+
+    private static class JoinBuilder {
+
+        private boolean removeGroupJoin;
+        private boolean selfJoin;
+        private boolean leftOuterJoin;
+        private boolean rightOuterJoin;
+        private boolean removeGroupFromIndex;
+        private boolean groupInWrongSpot;
+
+        private void build(RecordMetaDataBuilder builder) {
+            final JoinedRecordTypeBuilder joinBuilder = builder.addJoinedRecordType(JOIN_TYPE);
+            joinBuilder.addConstituent("parent", builder.getRecordType("MyParentRecord"), leftOuterJoin);
+            joinBuilder.addConstituent("child", builder.getRecordType("MyChildRecord"), rightOuterJoin);
+            if (!removeGroupJoin) {
+                joinBuilder.addJoin("parent", Key.Expressions.field("group"),
+                        selfJoin ? "parent" : "child", Key.Expressions.field("group"));
+            }
+            joinBuilder.addJoin("parent", Key.Expressions.field("child_rec_no"),
+                    "child", Key.Expressions.field("rec_no"));
+            final ThenKeyExpression concat;
+            if (removeGroupFromIndex) {
+                concat = Key.Expressions.concat(
+                        Key.Expressions.field("parent").nest(Key.Expressions.field("int_value")),
+                        Key.Expressions.field("child").nest(Key.Expressions.field("other_value")));
+            } else if (groupInWrongSpot) {
+                concat = Key.Expressions.concat(
+                        Key.Expressions.field("parent").nest(Key.Expressions.field("int_value")),
+                        Key.Expressions.field("parent").nest(Key.Expressions.field("group")),
+                        Key.Expressions.field("child").nest(Key.Expressions.field("other_value")));
+            } else {
+                concat = Key.Expressions.concat(
+                        Key.Expressions.field("parent").nest(Key.Expressions.field("group")),
```

Comment:
Is it worth adding a check that the first column could be `field("child").nest("group")` instead (that is, that using the first constituent isn't special)?